sigstore-java/src/main/java/dev/sigstore/bundle/BundleReader.java

```diff
@@ -0,0 +1,185 @@
+      if (dsseEnvelope.getSignaturesCount() != 1) {
+        throw new BundleParseException("DSEE envelopes must contain exactly one signature");
+      }
+      var dsseSignature =
+          ImmutableDSSESignature.builder()
+              .payload(dsseEnvelope.getPayload().toStringUtf8())
+              .payloadType(dsseEnvelope.getPayloadType())
+              .signature(dsseEnvelope.getSignatures(0).toByteArray())
+              .build();
+      bundleBuilder.dSSESignature(dsseSignature);
+    } else if (protoBundle.hasMessageSignature()) {
+      var signature = protoBundle.getMessageSignature().getSignature().toByteArray();
+      if (protoBundle.getMessageSignature().hasMessageDigest()) {
+        var hashAlgorithm = protoBundle.getMessageSignature().getMessageDigest().getAlgorithm();
+        if (hashAlgorithm != HashAlgorithm.SHA2_256) {
+          throw new BundleParseException(
+              "Cannot read message digests of type "
+                  + hashAlgorithm
+                  + ", only "
+                  + HashAlgorithm.SHA2_256
+                  + " is supported");
+        }
+        var messageSignature =
+            ImmutableMessageSignature.builder()
+                .messageDigest(
+                    ImmutableMessageDigest.builder()
+                        .hashAlgorithm(Bundle.HashAlgorithm.SHA2_256)
+                        .digest(
+                            protoBundle
+                                .getMessageSignature()
+                                .getMessageDigest()
+                                .getDigest()
+                                .toByteArray())
+                        .build())
+                .signature(signature)
+                .build();
+        bundleBuilder.messageSignature(messageSignature);
+      } else {
+        bundleBuilder.messageSignature(
+            ImmutableMessageSignature.builder().signature(signature).build());
+      }
+    } else {
+      throw new BundleParseException("A MessageSignature or DSSEEnvelope must be provided");
+    }
+
+    CertPath certPath;
+    try {
+      if (protoBundle.getVerificationMaterial().hasCertificate()) {
+        certPath =
+            ProtoMutators.toCertPath(
+                List.of(protoBundle.getVerificationMaterial().getCertificate()));
+      } else if (protoBundle.getVerificationMaterial().hasX509CertificateChain()) {
+        certPath =
+            ProtoMutators.toCertPath(
+                protoBundle
+                    .getVerificationMaterial()
+                    .getX509CertificateChain()
+                    .getCertificatesList());
+      } else if (protoBundle.getVerificationMaterial().hasPublicKey()) {
+        throw new BundleParseException("Plain public keys are not supported by this client");
+      } else {
+        throw new BundleParseException("Could not find a certificate or certificate chain");
+      }
+    } catch (CertificateException ce) {
```

Comment:
I'm not using an IDE, what throws ce in the above try block? Is it getVerificationMaterial?

Reply:
conversion to CertPath -- line 152